tests/test_utils.py

```diff
@@ -501,3 +502,171 @@ def test_gather_dimensions_exceptions():
+            utils_test_data.DA_R,
+            utils_test_data.DA_G,
+            "yellow",
+            None,
+            "blue",
+            utils.ERROR_SPECIFIED_NONPRESENT_REDUCE_DIMENSION2,
+        ),
+    ],
+)
+def test_gather_dimensions2_exceptions(
+    fcst, obs, weights, reduce_dims, preserve_dims, special_fcst_dims, error_msg_snippet
+):
+    """
+    Confirm `gather_dimensions2` raises exceptions as expected.
+    """
+    with pytest.raises(ValueError) as excinfo:
+        gather_dimensions2(
+            fcst,
+            obs,
+            weights=weights,
+            reduce_dims=reduce_dims,
+            preserve_dims=preserve_dims,
+            special_fcst_dims=special_fcst_dims,
+        )
+    assert error_msg_snippet in str(excinfo.value)
+
+
+def test_gather_dimensions2_warnings():
+    """Tests that gather_dimensions2 warns as expected with correct output."""
+    # Preserve "all" as a string but named dimension present in data
+    with pytest.warns(UserWarning):
+        result = gather_dimensions2(
+            utils_test_data.DA_R.rename({"red": "all"}), utils_test_data.DA_R, preserve_dims="all"
+        )
+        assert result == set([])
+
+    with pytest.warns(UserWarning):
+        result = gather_dimensions2(
+            utils_test_data.DA_R.rename({"red": "all"}), utils_test_data.DA_R, reduce_dims="all"
+        )
+        assert result == {"red", "all"}
+
+
+@pytest.mark.parametrize(
+    ("fcst", "obs", "weights", "reduce_dims", "preserve_dims", "special_fcst_dims", "expected"),
+    [
+        (utils_test_data.DA_B, utils_test_data.DA_R, None, None, None, None, {"blue", "red"}),
+        (utils_test_data.DA_B, utils_test_data.DA_R, utils_test_data.DA_G, None, None, None, {"blue", "red", "green"}),
+        (utils_test_data.DA_B, utils_test_data.DA_R, utils_test_data.DA_G, None, None, "blue", {"red", "green"}),
+        (utils_test_data.DA_B, utils_test_data.DA_R, utils_test_data.DA_G, None, None, ["blue"], {"red", "green"}),
+        (utils_test_data.DA_B, utils_test_data.DA_R, utils_test_data.DA_G, None, None, ["blue"], {"red", "green"}),
+        (utils_test_data.DA_B, utils_test_data.DA_R, utils_test_data.DA_G, "all", None, None, {"blue", "red", "green"}),
+        (utils_test_data.DA_B, utils_test_data.DA_R, utils_test_data.DA_G, "blue", None, None, {"blue"}),
+        (utils_test_data.DA_B, utils_test_data.DA_R, utils_test_data.DA_G, ["blue"], None, None, {"blue"}),
+        (utils_test_data.DA_RGB, utils_test_data.DA_R, utils_test_data.DA_G, ["green"], None, "blue", {"green"}),
+        (utils_test_data.DA_RGB, utils_test_data.DA_B, None, None, "all", "red", set([])),
+        (utils_test_data.DA_RGB, utils_test_data.DA_R, utils_test_data.DA_G, None, "green", None, {"red", "blue"}),
+        (utils_test_data.DA_RGB, utils_test_data.DA_R, None, None, ["green"], None, {"red", "blue"}),
+        (utils_test_data.DA_RGB, utils_test_data.DA_B, None, None, ["green"], "red", {"blue"}),
```

Comment:
Can you add a comment for each of these tests to be clearer what each of them are testing?

Reply:
done